Fix triggering post update actions after storing a URL Metric and refactor REST API endpoint logic into class

hbhalodia opened the pull request:
## Summary

<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
Fixes #1859 

## Relevant technical choices

<!-- Please describe your changes. -->
1. Converted `storage/rest-api.php` to class based approach and named `storage/class-od-rest-api.php`.
2. Converted the test file from `test-rest-api.php` to `test-class-rest-api.php`.
3. Updated the constants and functions which were directly used within files and used the static way to call the functions and class variables.

# TODO

- [x] Incorporate `add_action( 'od_trigger_page_cache_invalidation', 'od_trigger_page_cache_invalidation' )` after refactor.

<!--
For maintainers only, please make sure:

- PR has a `[Type]` label.
- PR has a plugin-specific milestone, or the `no milestone` label if it does not apply to any specific plugin.
- PR has a changelog-friendly title, or the `skip changelog` label if it should not be mentioned in the plugin's changelog.
-->

hbhalodia committed:
Add new class to encapsulate logic of rest api to make consitent codebase

This commit adds new class `class-od-rest-api.php` from `rest-api.php` file and adds the necessary required changes whereever normal functions were used, that are now refrencing the class static methods and variables. Also updated the test file test-rest-api.php to test-class-od-rest-api.php file and updated the required test cases as per the class.

Note: Have directly used function docs and comment that were present in the original file.

westonruter commented on plugins/image-prioritizer/helper.php:
```diff
@@ -298,7 +298,7 @@ function image_prioritizer_filter_rest_request_before_callbacks( $response, arra
 		||
 		// The strtolower() and outer trim are due to \WP_REST_Server::match_request_to_handler() using case-insensitive pattern match and using '$' instead of '\z'.
-		OD_REST_API_NAMESPACE . OD_URL_METRICS_ROUTE !== rtrim( strtolower( ltrim( $request->get_route(), '/' ) ) )
+		OD_Rest_API::OD_REST_API_NAMESPACE . OD_Rest_API::OD_URL_METRICS_ROUTE !== rtrim( strtolower( ltrim( $request->get_route(), '/' ) ) )
```
We'll have to be careful here. If someone updates Image Prioritizer first without also updating Optimization Detective, then they'll start getting fatal errors. The same would happen if they update Optimization Detective first without updating Image Prioritizer.

felixarntz replied:
We probably shouldn't make this change then, at least not yet. Unfortunately we can't catch when someone uses a constant to show a deprecation warning, but I wouldn't be too worried about other extensions outside of our own ones at this point.

So maybe we should just keep the global constants around a bit longer and add a `TODO` to remove them eventually, while already introducing the new ones and updating our extensions to use the new ones. It would probably be fine to remove the global constants in 1-2 months from now.

We could even put a big `Do not use them!` above in the code so that someone looking at the code doesn't get the wrong idea.

westonruter replied:
We can add `@deprecated` tags to the constants so they show up crossed-out when someone tries to use them.

westonruter commented on plugins/optimization-detective/storage/class-od-rest-api.php:
```diff
@@ -0,0 +1,358 @@
+ */
+class OD_Rest_API {
+
+	/**
+	 * Namespace for optimization-detective.
+	 *
+	 * @since 0.1.0
+	 * @access private
+	 * @var string
+	 */
+	const OD_REST_API_NAMESPACE = 'optimization-detective/v1';
+
+	/**
+	 * Route for storing a URL Metric.
+	 *
+	 * Note the `:store` art of the endpoint follows Google's guidance in AIP-136 for the use of the POST method in a way
+	 * that does not strictly follow the standard usage. Namely, submitting a POST request to this endpoint will either
+	 * create a new `od_url_metrics` post, or it will update an existing post if one already exists for the provided slug.
+	 *
+	 * @since 0.1.0
+	 * @access private
+	 * @link https://google.aip.dev/136
+	 * @var string
+	 */
+	const OD_URL_METRICS_ROUTE = '/url-metrics:store';
```
Maybe we should learn from the trouble we had the global constants with regard to there not being a way to indicate these being deprecated and change them to instead be methods like `get_namespace()` and `get_route()` that return the strings. This will give us more flexibility in the future.

hbhalodia replied:
Hi @westonruter, So I should revert the global constant what it was before `OD_REST_API_NAMESPACE ` and `OD_URL_METRICS_ROUTE ` and also add these 2 constants in class as well and then create a static function in class like `get_namespace()` and `get_route()` so that we can use that function inside `Optimization detective` and for `Image Priortizer` for now we should just use the global constant, so it does not break that.

Also, Can we add a check to `Image Priortizer` to first check if global constant exists, if yes use that if not, check for the `OD_Rest_API` class and call these 2 static methods to get this value?

westonruter replied:
I think these constants could be moved to a new `deprecated.php` in Optimization Detective and then they can continue to be used in Image Prioritizer for the time being. In a subsequent release we'll update Image Prioritizer to use what whatever new method we determine for accessing the namespace and route, such as methods on the new class.

westonruter replied:
We can also specify the specific minimum version of Optimization Detective that Image Prioritizer requires here:

https://github.com/WordPress/performance/blob/d21b74d7d12d4a378fe676a29c6e28ef1d0f9dae/plugins/image-prioritizer/helper.php#L24

But we can't go the other way around. So we have to keep those private deprecated constants in Optimization Detective around for awhile.

hbhalodia replied:
Okay, So creating a new method to access this from class is out of scope in this PR, correct?

So I also need to remove this constant from class as well for the time being and use global constant everywhere.

westonruter replied:
Within Optimization Detective itself we'll use the new method on the class. In fact, the methods can just return the constants for now. Eventually we'll get rid of the constants and have the new methods return strings directly.

hbhalodia replied:
Sounds Good, Would update the PR with required changes.

hbhalodia committed:
Update the class to create new methods to get the namespace and route for optimization detective

This commit adds the deprecated.php file which has the global constant that are being inside image priortizer and updated the OD plugin to use this constants via a dedicated methods.

hbhalodia commented:
Hi @westonruter, I have updated the PR with the required change, I need some inputs on,
1. I have used `@since n.e.x.t`, since I am not sure what exact version should be added.
2. Also, when I ran phpstan, it throws an error in image priortizer to use `OD_Rest_API` class methods to call those constant, as we have added a deprecated tag there, so how to remove that, for now I have committed using `--no-verify`.

Can you please suggest what should be there in both the cases?

Thank You,

hbhalodia committed:
Remove useless @return annotation

westonruter committed:
Remove class constants and address deprecated errors

westonruter commented:
> 1\. I have used `@since n.e.x.t`, since I am not sure what exact version should be added.

That is correct.

> 2\. Also, when I ran phpstan, it throws an error in image priortizer to use `OD_Rest_API` class methods to call those constant, as we have added a deprecated tag there, so how to remove that, for now I have committed using `--no-verify`.

Here's one way: bd3f309fd87c83f08bf87334b0c4ae03cde9e6ae.

westonruter commented on plugins/optimization-detective/storage/class-od-rest-api.php:
```diff
@@ -0,0 +1,360 @@
+	 * Registers endpoint for storage of URL Metric.
+	 *
+	 * @since 0.1.0
+	 * @access private
```
None of the methods in this class need `@access private` since the class as a whole is private.

westonruter replied:
That said, we'll want to eventually make the class non-private but make all methods except for `get_namespace()` and `get_route()` private. These are the only methods that other plugins may need, such as Image Prioritizer.

westonruter commented on plugins/optimization-detective/storage/class-od-rest-api.php:
```diff
@@ -0,0 +1,360 @@
+	/**
+	 * Registers endpoint for storage of URL Metric.
+	 *
+	 * @since 0.1.0
```
Probably should add a `@since n.e.x.t Moved from od_register_endpoint() function to class method`. Same for other methods in this class.

westonruter commented on plugins/optimization-detective/storage/class-od-rest-api.php:
```diff
@@ -0,0 +1,360 @@
+	 * @param string $origin Origin to check.
+	 * @return bool Whether the origin is allowed.
+	 */
+	private static function od_is_allowed_http_origin( string $origin ): bool {
```
The `od_` prefix should be removed from all methods.

hbhalodia commented on plugins/image-prioritizer/tests/test-helper.php:
```diff
@@ -727,7 +727,7 @@ public function data_provider_to_test_image_prioritizer_filter_rest_request_befo
 
 		$create_request = static function ( array $url_metric_data ): WP_REST_Request {
-			$request = new WP_REST_Request( 'POST', '/' . OD_REST_API_NAMESPACE . OD_URL_METRICS_ROUTE );
+			$request = new WP_REST_Request( 'POST', '/' . OD_Rest_API::get_namespace() . OD_Rest_API::get_route() );
```
Hi @westonruter, Would this not break, if we update this to use the `OD class` functions?

Though this is a test file, it won't work out of the box (not sure though).

westonruter replied:
In the tests we're fine to use the classes since it's all part of the same monorepo. The tests will never be run outside the context of this repo.

hbhalodia replied:
Yes, thought so. Thanks for the confirmation.

westonruter commented on plugins/optimization-detective/storage/class-od-rest-api.php:
```diff
@@ -0,0 +1,360 @@
+ * @since 0.1.0
+ * @access private
+ */
+class OD_Rest_API {
```
Instead of `OD_Rest_API` how about `OD_URL_Metric_Store_Endpoint`. Or maybe to follow core's naming convention more: `OD_REST_URL_Metric_Storage_Controller`. Let's see what @felixarntz thinks before you rename it.

hbhalodia replied:
Yes, would wait for the confirmation from @felixarntz.

hbhalodia committed:
Update the class to remove @access private on methods and removed od_ prefix from methods

hbhalodia commented:
Hi @westonruter, adding this for visibility, I would be going for long vacation starting tomorrow, so if we can land this PR before that, it would be very helpful, else I would not able to work on feedbacks after tomorrow 🙇. Apologies!

westonruter commented:
@hbhalodia got it! Unfortunately we may not be able to finish this before you go on vacation since my availability is limited now for the next week and a half per https://github.com/WordPress/performance/pull/1850#issuecomment-2654426506.

hbhalodia commented:
> @hbhalodia got it! Unfortunately we may not be able to finish this before you go on vacation since my availability is limited now for the next week and a half per [#1850 (comment)](https://github.com/WordPress/performance/pull/1850#issuecomment-2654426506).

Hey @westonruter, No issues. I will cover once back. Also, let's meet at WC Asia 🙌, I would be also there.

westonruter committed:
Merge branch 'trunk' of https://github.com/WordPress/performance into fix/issue-1859

westonruter commented:
@hbhalodia Did we end up crossing paths at WC Asia? I'm sorry to say I can't recall chatting with you, although there were so many people my memory may have been overloaded.

I've resolved the merge conflicts introduced by #1867 by porting the change to `rest-api.php` in that PR over to this class in 4f9b1c958.

hbhalodia committed:
Merge branch 'trunk' into fix/issue-1859

hbhalodia commented:
> @hbhalodia Did we end up crossing paths at WC Asia? I'm sorry to say I can't recall chatting with you, although there were so many people my memory may have been overloaded.
> 
> I've resolved the merge conflicts introduced by #1867 by porting the change to `rest-api.php` in that PR over to this class in [4f9b1c9](https://github.com/WordPress/performance/commit/4f9b1c95878b9cd64caf4b290fad67dc38f5f924).

Hi @westonruter, unfortunately not.

Thanks for resolving the conflicts.

Also is there anything pending in this PR other than, https://github.com/WordPress/performance/pull/1865#discussion_r1951993757?

Let me know as I am back from vacation and would work on the same.

Thank You,

westonruter commented:
@hbhalodia ah, sorry we couldn't connect in Manila.

For this PR, the bigger aspect to get clarity on is https://github.com/WordPress/performance/issues/1859#issuecomment-2691457161. Based on @felixarntz's feedback there we may not use a static class but rather a singleton class. You can expect a resolution to that discussion tomorrow.

hbhalodia commented:
> #1859 (comment)

Ah, alright, I missed that. No worries, we can update the PR based on the discussion and resolution.

Thank You,

westonruter commented:
@hbhalodia OK, I chatted with @felixarntz and we need to discuss more tomorrow, but in the meantime I think we've agreed that the class should not be static but rather be instantiated. The remaining question then is whether we implement it as a singleton or not. In the meantime, however, you could put it in `hooks.php` as:

```php
( new OD_Whatever_We_Call_This_Class() )->add_hooks();
```

However, read on...

As for the naming, we should probably follow the example of the REST API classes used in core: https://github.com/WordPress/wordpress-develop/tree/trunk/src/wp-includes/rest-api/endpoints

So perhaps `OD_REST_URL_Metrics_Store_Controller`.

Maybe for one specific endpoint which could be used as a model: https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-edit-site-export-controller.php

The class could extend `WP_REST_Controller` in the same way, with the additional static methods for `get_namespace()` and `get_route()` (although perhaps `get_rest_base()` would then be more appropriate now).

It gets a bit tricky here because endpoints typically are for resources (nouns) whereas this endpoint is exclusively for an action to store the provided URL Metric. This is reflected in the endpoint name including a colon (`url-metrics:store`) with this comment:

https://github.com/WordPress/performance/blob/db322a37e675aeea54db7afba8a31ac0c40e6393/plugins/optimization-detective/storage/rest-api.php#L25-L29

Note that `WP_REST_Edit_Site_Export_Controller` uses `wp-block-editor/v1/export` as its endpoint. And is `export` here a noun or a verb? Maybe a noun since it only implements a `GET` method, and yet our endpoint only implements `POST`.

If extending `WP_REST_Controller`, then I don't think we should have an `add_hooks()` method on it. Instead, there could be an `od_create_initial_rest_routes()` function (cf. [`create_initial_rest_routes()`](https://github.com/WordPress/wordpress-develop/blob/ac057e6c692e7e88b1758eb8dc6dee2312bb5c7b/src/wp-includes/rest-api.php#L260-L419) in core) which is added to `detection.php`. For example:

```php
/**
 * Registers REST API routes.
 *
 * @since n.e.x.t
 */
function od_create_initial_rest_routes() {
	$controller = new OD_REST_URL_Metrics_Store_Controller();
	$controller->register_routes();
}
```

And in `hooks.php` it would do:

```php
add_action( 'rest_api_init', 'od_create_initial_rest_routes' );
```

This would follow most closely the patterns used in core, I believe.

hbhalodia commented:
Hi @westonruter, thanks for the detailed comment, I had a few questions. As you have shared the 2 approach here, either to use class to add the action and register endpoint, or use the function add it to `detection.php` and in that function instantiate the class and add the routes.

IMO, if we need to make it consistent for now with the codebase, we should follow the first approach as of now, to remove static method and use add_hooks() class function to register the action, and also not needed to extend the rest controller class.

I am adding the first approach for now, we can discuss and can change in this PR later once you provide a confirmation.

Thank You,

hbhalodia committed:
Update the class to use specific naming convention and remove static class method to add hooks

hbhalodia committed:
Merge branch 'trunk' into fix/issue-1859

felixarntz commented:
@westonruter @hbhalodia Following up on your above comment https://github.com/WordPress/performance/pull/1865#issuecomment-2699453255 with a few more thoughts:

> The class could extend `WP_REST_Controller` in the same way, with the additional static methods for `get_namespace()` and `get_route()` (although perhaps `get_rest_base()` would then be more appropriate now).

It's worth noting that the Core base class is not particularly great in design when considering it as a starting point for a _general_ REST endpoint. The `WP_REST_Controller` class works well for a set of _multiple_ endpoints that deal with a specific resource, i.e. offering CRUD operations for it. But it's a poor abstraction when all we want to do is implement a single endpoint with a specific purpose like we're doing here. You can tell from how many Core classes that extend `WP_REST_Controller` have to contain empty methods just because the parent class requires them.

You indirectly mention this as well via:
> It gets a bit tricky here because endpoints typically are for resources (nouns) whereas this endpoint is exclusively for an action to store the provided URL Metric.

So from that perspective, I think it would be better to implement a custom class (not extending `WP_REST_Controller`), which we could call something like `OD_REST_URL_Metrics_Store_Endpoint` - basically the same name as proposed above, but focusing on the single endpoint it implements. We could take over some of the method signatures from `WP_REST_Controller` where applicable, but we wouldn't need to handle all the bloat it would add that would be without a purpose in our class.

Of course, WordPress Core uses `WP_REST_Controller` for all its REST endpoint related classes, so if we want to align with that for the sake of consistency, I'm not strongly opposed. But worth pointing out that that's not a good call from an API design perspective other than for consistency.

> If extending `WP_REST_Controller`, then I don't think we should have an `add_hooks()` method on it. Instead, there could be an `od_create_initial_rest_routes()` function (cf. [`create_initial_rest_routes()`](https://github.com/WordPress/wordpress-develop/blob/ac057e6c692e7e88b1758eb8dc6dee2312bb5c7b/src/wp-includes/rest-api.php#L260-L419) in core) which is added to `detection.php`.

This sounds good to me, though I would say we can use that approach even if we don't extend `WP_REST_Controller`, but go with a custom endpoint class as I mentioned above. Since you're already mentioning to not have an `add_hooks()` method on the class, I think even better then would be to decouple registration from that class entirely and instead just have getter methods like `get_namespace()`, `get_route()`, and `get_registration_args()`, so that the class is only responsible for providing the endpoint definition, but registration is handled entirely via that action hook outside of the class.

If it still has registration baked in, I think we might as well include the `add_hooks()` to have all aspects of registration baked in - but I tend to agree with you that leaving that part out of the class is a cleaner implementation, as it would for example make it very easy to later implement more endpoint classes and reuse the same registration logic.

hbhalodia committed:
Merge branch 'trunk' into fix/issue-1859

hbhalodia committed:
Update the class to remove add_hook and use action outside class to register route

Added the test case as well for the new function added in detection.php and call the action to register route and use class independently just for route registration

hbhalodia commented:
Hi @felixarntz @westonruter, I have update the PR with the required changes as mentioned in this commit - https://github.com/WordPress/performance/pull/1865/commits/fb1a5a10e33a5b7758ee0e1abcac82c4f9516a13, Can you please review if this is what we are expecting here.

Thank You,

felixarntz commented on plugins/image-prioritizer/helper.php:
```diff
@@ -298,7 +298,11 @@ function image_prioritizer_filter_rest_request_before_callbacks( $response, arra
+			OD_REST_API_NAMESPACE . OD_URL_METRICS_ROUTE // @phpstan-ignore constant.deprecated, constant.deprecated (To be replaced with class method calls in subsequent release.)
+			!==
+			rtrim( strtolower( ltrim( $request->get_route(), '/' ) ) )
+		)
```
In addition to this annotation, we should probably have conditional logic here to rely on the new class methods if the `class_exists`? That way the Image Prioritizer implementation remains cross-version compatible before and after this change in Optimization Detective.

hbhalodia replied:
Sure, I would add the conditional logic here to use the class methods if class exists or use the so called deprecated constants.

felixarntz commented on plugins/optimization-detective/deprecated.php:
```diff
@@ -0,0 +1,39 @@
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Namespace for optimization-detective.
+ *
+ * @since 0.1.0
+ * @access private
+ * @deprecated n.e.x.t This constant should not be used, instead use OD_REST_URL_Metrics_Store_Controller::get_namespace() method.
+ * @var string
+ */
+const OD_REST_API_NAMESPACE = 'optimization-detective/v1';
+
+/**
+ * Route for storing a URL Metric.
+ *
+ * Note the `:store` art of the endpoint follows Google's guidance in AIP-136 for the use of the POST method in a way
+ * that does not strictly follow the standard usage. Namely, submitting a POST request to this endpoint will either
+ * create a new `od_url_metrics` post, or it will update an existing post if one already exists for the provided slug.
+ *
+ * @since 0.1.0
+ * @access private
+ * @link https://google.aip.dev/136
+ * @deprecated n.e.x.t This constant should not be used, instead use OD_REST_URL_Metrics_Store_Controller::get_route() method.
+ * @var string
+ */
+const OD_URL_METRICS_ROUTE = '/url-metrics:store';
```
We should probably open a follow up issue already to remove these constants in 1-2 releases later, just so that we don't forget :)

hbhalodia replied:
Should we open the issue now? or once the PR gets merged? since currently, we do not have any proper link for this we can add in issue description.

felixarntz replied:
We can open the issue at any time, feel free to already do so. We can always include a better link later.

hbhalodia replied:
Hi @felixarntz, I have created the issue so that we can keep track to resolve once we have newer releases. https://github.com/WordPress/performance/issues/1920

Let me know if anything is needed here to add/remove from the issue.

felixarntz commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-controller.php:
```diff
@@ -0,0 +1,359 @@
+ *
+ * @access private
+ */
+class OD_REST_URL_Metrics_Store_Controller {
```
The class implementation now is almost entirely static, so that's contrary to what was discussed on the PR before, I think.

Per https://github.com/WordPress/performance/pull/1865#issuecomment-2701757708, we should rather call this class `OD_REST_URL_Metrics_Store_Endpoint`, and registration should not be part of the class, instead it should have a getter method like `get_registration_args()`.

felixarntz commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-controller.php:
```diff
@@ -0,0 +1,359 @@
+	 *
+	 * @see od_compose_site_health_result()
+	 */
+	public function register_endpoint(): void {
```
See above, per https://github.com/WordPress/performance/pull/1865#issuecomment-2701757708 this method should not be included in the class. Instead the class should have a method (e.g. `get_registration_args()`) to return the data for the third `register_rest_route` parameter as an array.

felixarntz reviewed (changes requested):
@hbhalodia Thank you for working on this! The logic looks good, though I think the PR still require some updates based on the previous discussion. See my inline feedback below.

hbhalodia committed:
Add the conditional check for class_exists and use constant or class method accordingly

hbhalodia committed:
Update the class name to use OD_REST_URL_Metrics_Store_Endpoint

hbhalodia committed:
Update the class method to register the args and add getter function

hbhalodia commented:
Hi @felixarntz @westonruter, I have updated the PR with the requested changes.

Also, Since we have remove the route registration from the class, and have added the getter method to get the args and use the register_rest_route outside of the class to register the endpoint. I have modified the test and remove the covers `register_endpoint` from test class.

Let me know your feedbacks on the newer commit in the PR and would update the same.

**Note**: We have some failing tests, but they are not from this PR, I tried to checkout to trunk and run the test, it still shows the test being failed. So not sure we should include that test in this PR.

Thank You,

westonruter committed:
Merge branch 'trunk' into fix/issue-1859

westonruter commented:
> **Note**: We have some failing tests, but they are not from this PR, I tried to checkout to trunk and run the test, it still shows the test being failed. So not sure we should include that test in this PR.

Sorry for the confusion there. The test failure was fixed in #1913, which I've now added to this PR by merging in `trunk`.

westonruter commented on plugins/optimization-detective/detection.php:
```diff
@@ -157,3 +157,23 @@ static function ( OD_URL_Metric_Group $group ): array {
+/**
+ * Create the rest request and register the endpoint routes for the URL Metrics Store.
+ *
+ * @since n.e.x.t
```
```suggestion
 * @since n.e.x.t
 * @access private
```

westonruter commented on plugins/optimization-detective/detection.php:
```diff
@@ -157,3 +157,23 @@ static function ( OD_URL_Metric_Group $group ): array {
 }
+
+/**
+ * Create the rest request and register the endpoint routes for the URL Metrics Store.
```
```suggestion
 * Registers the REST API endpoint for storing URL Metrics.
```

westonruter commented on plugins/optimization-detective/detection.php:
```diff
@@ -157,3 +157,23 @@ static function ( OD_URL_Metric_Group $group ): array {
+ *
+ * @since n.e.x.t
+ */
+function od_create_rest_url_metric_store_rest_request(): void {
```
```suggestion
function od_register_rest_url_metric_store_endpoint(): void {
```

westonruter commented on plugins/optimization-detective/hooks.php:
```diff
@@ -24,4 +24,5 @@
 add_filter( 'site_status_tests', 'od_add_rest_api_availability_test' );
 add_action( 'admin_init', 'od_maybe_run_rest_api_health_check' );
 add_action( 'after_plugin_row_meta', 'od_render_rest_api_health_check_admin_notice_in_plugin_row', 30 );
+add_action( 'rest_api_init', 'od_create_rest_url_metric_store_rest_request' );
```
```suggestion
add_action( 'rest_api_init', 'od_register_rest_url_metric_store_endpoint' );
```

westonruter commented on plugins/optimization-detective/detection.php:
```diff
@@ -157,3 +157,23 @@ static function ( OD_URL_Metric_Group $group ): array {
+function od_create_rest_url_metric_store_rest_request(): void {
+
+	$endpoint_controller = new OD_REST_URL_Metrics_Store_Endpoint();
+	$route_args          = $endpoint_controller->get_registration_args();
+
+	/**
+	 * Register the endpoint.
+	 */
+	register_rest_route(
+		$endpoint_controller::get_namespace(),
+		$endpoint_controller::get_route(),
+		$route_args
+	);
```
```suggestion
	register_rest_route(
		$endpoint_controller::get_namespace(),
		$endpoint_controller::get_route(),
		$endpoint_controller->get_registration_args()
	);
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+ * OD_REST_URL_Metrics_Store_Endpoint class
+ *
+ * @since n.e.x.t
+ *
```
```suggestion
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+class OD_REST_URL_Metrics_Store_Endpoint {
+
+	/**
+	 * Get the namespace for optimization-detective
```
```suggestion
	 * Gets the namespace for the REST API endpoint.
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	/**
+	 * Get the namespace for optimization-detective
+	 *
+	 * @since n.e.x.t
```
```suggestion
	 * @since n.e.x.t
	 *
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	 * create a new `od_url_metrics` post, or it will update an existing post if one already exists for the provided slug.
+	 *
+	 * @since n.e.x.t
+	 *
```
```suggestion
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+ *
+ * @since n.e.x.t
+ *
+ * @access private
```
Perhaps this should not be private because we are intending for the `OD_REST_URL_Metrics_Store_Endpoint::get_namespace()` and `OD_REST_URL_Metrics_Store_Endpoint::get_route()` static methods to be used by other plugins? Nevertheless, the fact that Image Prioritizer is directly accessing the endpoint is itself an exception, a workaround implemented in https://github.com/WordPress/performance/pull/1713 when I was originally intending for there to be an explicit API for doing deeper validation of submitted URL Metrics prior to storage.

hbhalodia replied:
So should we remove the `@access private` from the class? But here we are intending to remove the static method and making it constant - https://github.com/WordPress/performance/pull/1865#discussion_r1990292303

westonruter replied:
Since the constant is (kinda) intended to be used by other plugins, then `@access private` should be removed from the class but then added to all of the methods (but omitted from these two constants).

hbhalodia replied:
Sounds good, Would update the PR.

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	}
+
+	/**
+	 * Get the args for the URL Metric storage endpoint.
```
```suggestion
	 * Gets the arguments for registering the endpoint.
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+			),
+		);
+
+		/**
+		 * Create route args.
+		 */
+		$route_args = array(
```
```suggestion
		return array(
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+				}
+				return true;
+			},
+		);
+
+		return $route_args;
```
```suggestion
		);
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	 * @see is_allowed_http_origin()
+	 *
+	 * @param string $origin Origin to check.
+	 *
```
```suggestion
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	 * the `is_allowed_http_origin()` function in core for some reason returns a string rather than a boolean.
+	 *
+	 * @since n.e.x.t
+	 *
```
```suggestion
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	 * @phpstan-param WP_REST_Request<array<string, mixed>> $request
+	 *
+	 * @param WP_REST_Request $request Request.
+	 *
```
```suggestion
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+				$args,
+				rest_get_endpoint_args_for_schema( OD_Strict_URL_Metric::get_json_schema() )
+			),
+			'callback'            => static function ( WP_REST_Request $request ) {
+				return self::handle_rest_request( $request );
+			},
```
```suggestion
			'callback'            => array( $this, 'handle_rest_request' ),
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	 *
+	 * @return WP_REST_Response|WP_Error Response.
+	 */
+	private static function handle_rest_request( WP_REST_Request $request ) {
```
```suggestion
	public function handle_rest_request( WP_REST_Request $request ) {
```

felixarntz replied:
+1, this shouldn't be static.

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	 *
+	 * @return bool Whether the origin is allowed.
+	 */
+	private static function is_allowed_http_origin( string $origin ): bool {
```
```suggestion
	protected static function is_allowed_http_origin( string $origin ): bool {
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	}
+
+	/**
+	 * Handles REST API request to store metrics.
```
```suggestion
	 * Handles the REST API request to store a URL Metric.
```

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	 *
+	 * @param positive-int $cache_purge_post_id Cache purge post ID.
+	 */
+	public static function trigger_page_cache_invalidation( int $cache_purge_post_id ): void {
```
Huh, I don't see anywhere that this function is actually being used in the codebase, outside of the tests.

Previously this method was in the `od_trigger_page_cache_invalidation()` function which at a glance would seem like it was before used here:

https://github.com/WordPress/performance/blob/2e2749d22b054a32daa9cec1cbbb023da84b418d/plugins/optimization-detective/storage/rest-api.php#L269-L277

But now with the method change it becomes clear that this was never getting called at all! The `od_trigger_page_cache_invalidation` here is the action name, _not_ the function name. In reality, the action name is a "trigger for the trigger" to clear the cache.

What was missing was something that did:

```
add_action( 'od_trigger_page_cache_invalidation', 'od_trigger_page_cache_invalidation' );
```

However, with the refactor this would need to be:

```
add_action( 'od_trigger_page_cache_invalidation', array( OD_REST_URL_Metrics_Store_Endpoint::class, 'trigger_page_cache_invalidation' ) );
```

But it seems like this shouldn't be static. So maybe inside of ~`od_create_rest_url_metric_store_rest_request()`~ `od_register_rest_url_metric_store_endpoint()`  it can do:

```php
register_rest_route(
	$endpoint_controller::get_namespace(),
	$endpoint_controller::get_route(),
	$endpoint_controller->get_registration_args()
);
add_action( 'od_trigger_page_cache_invalidation', array( $endpoint_controller, 'trigger_page_cache_invalidation' ) );
```

Or maybe it would make sense to keep this method outside as a separate function because I can imagine some sites wanting to be able to do `remove_action()` on this function if they don't want to enable storing URL Metrics to invalidate page caches (which apparently is the current case, by accident!)

westonruter replied:
I can confirm that page cache invalidation in `trunk` is currently broken. It is fixed by adding this to `hooks.php`:

```php
add_action( 'od_trigger_page_cache_invalidation', 'od_trigger_page_cache_invalidation' );
```

So this will need to be accounted for here.

felixarntz replied:
+1, this shouldn't be static. It could be hooked by using an instance method like you're saying above.

hbhalodia replied:
I would create a different callback to add this from hook.php.

>  I can imagine some sites wanting to be able to do remove_action() on this function if they don't want to enable storing URL Metrics to invalidate page caches (which apparently is the current case, by accident!)

This would make sure the above point as well.

hbhalodia replied:
I have added this here - Let me know if this works well - https://github.com/WordPress/performance/blob/a87d3563e26f4ec0ca0a7ab848b9ec31dd42d7c7/plugins/optimization-detective/detection.php#L178-L189 and called from hooks.php - https://github.com/WordPress/performance/blob/a87d3563e26f4ec0ca0a7ab848b9ec31dd42d7c7/plugins/optimization-detective/hooks.php#L28 Thank You,

westonruter replied:
I think the method should be removed from the class entirely and instead put into this `od_trigger_page_cache_invalidation_callback()` function.

And let's call the function `od_handle_trigger_page_cache_invalidation()` instead, maybe.

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+ *
+ * @access private
+ */
+class OD_REST_URL_Metrics_Store_Endpoint {
```
```suggestion
final class OD_REST_URL_Metrics_Store_Endpoint {
```

westonruter commented on plugins/optimization-detective/tests/test-detection.php:
```diff
@@ -225,4 +225,27 @@ public function test_od_get_detection_script_returns_script( Closure $set_up, ar
 	}
+
+	/**
+	 * Test od_create_rest_url_metric_store_rest_request().
+	 *
+	 * @covers ::od_create_rest_url_metric_store_rest_request
+	 */
+	public function test_od_create_rest_url_metric_store_rest_request(): void {
```
```suggestion
	 * Test od_register_rest_url_metric_store_endpoint().
	 *
	 * @covers ::od_register_rest_url_metric_store_endpoint
	 */
	public function test_od_register_rest_url_metric_store_endpoint(): void {
```

westonruter commented on plugins/optimization-detective/tests/test-detection.php:
```diff
@@ -225,4 +225,27 @@ public function test_od_get_detection_script_returns_script( Closure $set_up, ar
+	 */
+	public function test_od_create_rest_url_metric_store_rest_request(): void {
+
+		global $wp_rest_server;
+
+		if ( ! isset( $wp_rest_server ) ) {
+			$wp_rest_server = new WP_REST_Server();
+
+			// Run the action to register the routes.
+			do_action( 'rest_api_init' );
+		}
+
+		// Get the routes.
+		$routes = $wp_rest_server->get_routes();
```
This should rather do the following:

```suggestion
		$routes = rest_get_server()->get_routes();
```

If WordPress doesn't clean up the `$wp_rest_server` global then we should do it in `set_up`/`tear_down`, for example, do `unset( $GLOBALS['wp_rest_server'] );`.

felixarntz commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	 * @since n.e.x.t
+	 * @return non-empty-string Namespace.
+	 */
+	public static function get_namespace(): string {
```
Seeing this, I'm not sure this is how we'll want to implement this. For easy access without instantiation, it makes sense to have it as a static method, but since it just returns a constant string without any logic involved, I think it would be more suitable to make it a `const` on the class.

That way we furthermore avoid having a mix of instance and static methods on this class, but most importantly I don't see any benefits of using a static method over `const`.

westonruter replied:
Actually, the benefit is why these were moved from global constants to static methods in the first place: methods can be deprecated with warnings but constants cannot. See:

* https://github.com/WordPress/performance/pull/1865#discussion_r1951614840
* https://github.com/WordPress/performance/pull/1865#discussion_r1951715603

felixarntz replied:
Hmm, not sure I agree that's a good reason to have a static method. Given that philosophy, constants shouldn't exist.

Ideally, constants don't need to be deprecated because their point is to maintain a constant (ha!) association to whatever their value is. It is expected for a constant to change its value, and as long as the constant is consistently used (instead of the underlying value), it's fine.

westonruter replied:
I see that the endpoint controller classes in core use constants, for example:

* https://github.com/WordPress/wordpress-develop/blob/7475126055ddf7b9fd6c303d2468676926b762f0/src/wp-includes/rest-api/endpoints/class-wp-rest-search-controller.php#L19-L47
* https://github.com/WordPress/wordpress-develop/blob/7475126055ddf7b9fd6c303d2468676926b762f0/src/wp-includes/rest-api/endpoints/class-wp-rest-font-faces-controller.php#L15-L21
* https://github.com/WordPress/wordpress-develop/blob/7475126055ddf7b9fd6c303d2468676926b762f0/src/wp-includes/rest-api/endpoints/class-wp-rest-plugins-controller.php#L19

But they don't have any static methods, so let's go ahead and use constants. Let's just be sure we give them them the right name so we don't ever want to change them 😄

hbhalodia replied:
> But they don't have any static methods, so let's go ahead and use constants. Let's just be sure we give them them the right name so we don't ever want to change them 😄

Have converted the methods to constants, and provided the name as `REST_API_NAMESPACE` and `REST_API_ROUTE`.

Let me know if we need to change this.

felixarntz replied:
WDYT about shortening them to just `REST_NAMESPACE` and `REST_ROUTE`?

westonruter replied:
Per https://github.com/WordPress/performance/pull/1865#discussion_r1991637846 and https://github.com/WordPress/performance/pull/1865#discussion_r1991638943 I think they should be just `NAMESPACE` and `ROUTE`. The "`REST`" part is redundant since it is already part of the class name.

felixarntz replied:
Works for me, though with single words like this I'm sometimes unsure if you run into conflicts with PHP reserved terms (e.g. relevant for `NAMESPACE`). A prefix generally helps avoid such conflicts.

westonruter replied:
Looks like it's not a [reserved word](https://www.php.net/manual/en/reserved.other-reserved-words.php), at least since 7.0: https://3v4l.org/P5mWG

(Interestingly, some versions of 5.x have problems with this but not others.)

felixarntz replied:
I'm still thinking in principle using a prefix is a safer pattern from a future compat perspective, should we want to add more class constants. For example we could use `ROUTE_NAMESPACE` and `ROUTE_BASE`, which IMO would work well because technically both of them together make up the actual route path (and the word "base" is also used by Core).

But I don't feel strongly about it, either way works.

westonruter replied:
A `NAMESPACE` class constant is also being used by the `woocommerce-paypal-payments`, `official-facebook-pixel`, `pixelyoursite`, `give`, and `wp-job-manager` plugins: https://wpdirectory.net/search/01JP5GSDAP0ETEFRGE9QK5J9F0

westonruter replied:
But `ROUTE_NAMESPACE` and `ROUTE_BASE` also work for me.

hbhalodia replied:
Converting to `ROUTE_NAMESPACE` and `ROUTE_BASE` 👍.

felixarntz commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,364 @@
+	 *
+	 * @return non-empty-string Route.
+	 */
+	public static function get_route(): string {
```
Same here, I think this would be better a `const` on the class.

westonruter replied:
See https://github.com/WordPress/performance/pull/1865#discussion_r1990209610

felixarntz commented on plugins/optimization-detective/tests/test-site-health.php:
```diff
@@ -420,7 +420,7 @@ protected function filter_rest_api_response( $mocked_response ): object {
 			'pre_http_request',
 			static function ( $pre, array $args, string $url ) use ( $mocked_response, $observer ) {
-				if ( rest_url( OD_REST_API_NAMESPACE . OD_URL_METRICS_ROUTE ) === $url ) {
+				if ( rest_url( OD_REST_URL_Metrics_Store_Endpoint::get_namespace() . OD_REST_URL_Metrics_Store_Endpoint::get_route() ) === $url ) {
```
See notes above on using a `const`.

felixarntz commented on plugins/optimization-detective/detection.php:
```diff
@@ -157,3 +157,23 @@ static function ( OD_URL_Metric_Group $group ): array {
+	 * Register the endpoint.
+	 */
+	register_rest_route(
+		$endpoint_controller::get_namespace(),
+		$endpoint_controller::get_route(),
```
Maybe a personal preference, but feels a bit strange to invoke a static method on an instance. But also per my other comment, I think we can replace this with a const.

IMO, something like `OD_REST_URL_Metrics_Store_Endpoint::NAMESPACE` is easier to understand than `$endpoint_controller::NAMESPACE`.

felixarntz reviewed (commented):
There are still too many static methods on the class that don't need to be static. We should avoid having a mix of static and non-static methods, except where static methods make sense (a stateless method).

For the namespace and route, I think using `const` would be a better solution.

hbhalodia committed:
Fix the error feedbacks

1. Related to comments and common naming convention.

hbhalodia committed:
Merge branch 'fix/issue-1859' of github.com:hbhalodia/performance into fix/issue-1859

hbhalodia committed:
Merge branch 'trunk' into fix/issue-1859

hbhalodia committed:
Remove the static method from the class and updated to use the constant

hbhalodia commented:
Hi @westonruter @felixarntz, I have updated the PR with the requested changes. Can you please check this unresolved comments and provide your feedback on the same?

Also, I have added a new function callback for `page cache invalidation` in `detection.php`, but not have added the test for it in `test-detection.php`, because this is already covered in https://github.com/WordPress/performance/blob/a87d3563e26f4ec0ca0a7ab848b9ec31dd42d7c7/plugins/optimization-detective/tests/storage/test-class-od-rest-url-metrics-store-endpoint.php#L849, So I have just added the `@covers ::function_name`.

Let me know if this could work, or should I add the same test function in `test-detection.php` file as well.

Thank You,

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,347 @@
+ * @since n.e.x.t
+ * @access private
+ */
+final class OD_REST_URL_Metrics_Store_Endpoint {
```
@felixarntz just seeing this plural "Metrics" here and I'm now second guessing myself on this name. The endpoint is `url-metrics:store` so this is consistent, but in reality the endpoint takes a single URL Metric to store (in an `od_url_metrics` post which _is_ a collector multiple). Other classes like `OD_URL_Metric_Group_Collection` are singular since the "group" and "collection" indicate the plurality. So should the endpoint actually be `url-metric:store` and the class name updated to match here?

felixarntz replied:
I think that would make sense, though I don't have a strong preference.

westonruter replied:
After consulting with Gemini, it argues that `/url-metrics:store` is more appropriate:

<blockquote>
According to AIP-122, collection IDs should be plural nouns. While your custom method isn't a collection ID itself, it operates *in the context of* the conceptual collection of metrics for a given URL.

Even though the request body to your custom method contains a single metric, the *intent* of the operation is to store that metric within the set of metrics associated with a URL. The "store" action is happening relative to the `url-metrics` for a specific URL.

Therefore, I would argue that **`/url-metrics:store` is more appropriate**.

Here's why:

1.  **Conceptual Target:** The custom method's logic is primarily concerned with the `UrlMetrics` resource (plural) for a given URL. It either creates this resource or modifies its `metrics` collection.
2.  **Consistency:** If you were to add other custom methods related to the `UrlMetrics` resource (e.g., maybe a method to get a summary, `/url-metrics:summarize`), it maintains a consistent naming scheme.
3.  **AIP-122:** While not directly a collection ID, the custom method name benefits from aligning with the idea that it's operating on a collection-like entity (`url-metrics` for a URL).

Think of it like this: you're not just storing "a URL metric" in isolation; you're storing it *within the context of* the "URL metrics" for a particular URL.

So, stick with the plural form: `/url-metrics:store`.
</blockquote>

So let's stick with `/url-metrics:store` for the REST API method. Also, when I said that all the other classes are singular this isn't actually true. There's also an `OD_URL_Metrics_Post_Type` class. So after these considerations, I think we should stick with `OD_REST_URL_Metrics_Store_Endpoint`.

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,347 @@
+	 * @since n.e.x.t
+	 * @var string
+	 */
+	const REST_API_NAMESPACE = 'optimization-detective/v1';
```
Should this just be:

```suggestion
	const NAMESPACE = 'optimization-detective/v1';
```

hbhalodia replied:
Update with - https://github.com/WordPress/performance/pull/1865#discussion_r1991808494

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,347 @@
+	 * @link https://google.aip.dev/136
+	 * @var string
+	 */
+	const REST_API_ROUTE = '/url-metrics:store';
```
Likewise should this just be:

```suggestion
	const ROUTE = '/url-metrics:store';
```

Otherwise it is redundant.

hbhalodia replied:
Resolved with - https://github.com/WordPress/performance/pull/1865#discussion_r1991808494

westonruter commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,347 @@
+				rest_get_endpoint_args_for_schema( OD_Strict_URL_Metric::get_json_schema() )
+			),
+			'callback'            => array( $this, 'handle_rest_request' ),
+			'permission_callback' => static function () {
```
This maybe should be moved into a class method as well, instead of a closure.

hbhalodia committed:
Update the trigger page cache validation function to remove from class and add it separate method

hbhalodia committed:
Merge branch 'trunk' into fix/issue-1859

hbhalodia committed:
Update the class constants to use new names

hbhalodia committed:
Add the method for permission callback instead of closure

hbhalodia committed:
Remove access private from class and add to individual methods other than constant

westonruter commented:
Sorry, after merging https://github.com/WordPress/performance/pull/1905 there are now some merge conflicts that need to be resolved. I think the new function which was introduced in that PR should made a method of the class introduced here

westonruter commented:
I'm working on resolving the merge conflicts.

westonruter committed:
Fix merge conflicts with #1905

westonruter commented:
That was more difficult than I was expecting.

westonruter committed:
Rename od_handle_trigger_page_cache_invalidation() to od_trigger_post_update_actions()

westonruter committed:
Address minor static analysis complaints

westonruter committed:
Improve permission check method name and fix test case name

westonruter committed:
Remove obsolete class aliases deprecated in October 2024 via #1574

westonruter committed:
Add codeCoverageIgnore

westonruter committed:
Double-check that od_trigger_post_update_actions() is getting called

westonruter commented on plugins/optimization-detective/load.php:
```diff
@@ -107,14 +110,12 @@ static function ( string $version ): void {
 		require_once __DIR__ . '/class-od-strict-url-metric.php';
 		require_once __DIR__ . '/class-od-url-metric-group.php';
 		require_once __DIR__ . '/class-od-url-metric-group-collection.php';
-		class_alias( OD_URL_Metric_Group::class, 'OD_URL_Metrics_Group' ); // Temporary class alias for back-compat after rename.
-		class_alias( OD_URL_Metric_Group_Collection::class, 'OD_URL_Metrics_Group_Collection' ); // Temporary class alias for back-compat after rename.
```
I removed these because their removal was long-overdue, as these were deprecated in October 2024.

See also this deprecated accessor which can be removed soon as well:

https://github.com/WordPress/performance/blob/da320f3705a95a379993451d30818c379486aaad/plugins/optimization-detective/class-od-tag-visitor-context.php#L123-L136

westonruter replied:
Now tracking as part of https://github.com/WordPress/performance/issues/1920

westonruter committed:
Move class aliases to deprecated.php

westonruter reviewed (approved):
@hbhalodia Thank you for the many iterations on this!

westonruter committed:
Revert moving class aliases to deprecated.php for sake of PHPStan

felixarntz commented on plugins/optimization-detective/storage/class-od-rest-url-metrics-store-endpoint.php:
```diff
@@ -0,0 +1,371 @@
+	 * @param WP_REST_Request $request Request used to generate the response.
+	 * @return mixed Passed through $result if successful, or otherwise a WP_Error.
+	 */
+	public function decompress_rest_request_body( $result, WP_REST_Server $server, WP_REST_Request $request ) {
```
Having this in here doesn't seem ideal, since the method is not scoped to this endpoint (even though it technically only acts if the endpoint is hit).

So I think ideally this would be either a standalone function or in a general REST utility class with static methods.

Since it's marked private, not a blocker for merging, but wanted to flag. It could be refactored later.

westonruter replied:
OK, I'll address in a follow-up I'm working on right now.

westonruter replied:
See https://github.com/WordPress/performance/pull/1924

felixarntz reviewed (approved):
Thanks @hbhalodia and @westonruter!

This looks good, just one point of feedback, but not a blocker for merge, though it would be great to follow up on this separately.

westonruter commented:
I can see this is working now because in my Stream records on my site I see that the posts are updated:

![image](https://github.com/user-attachments/assets/9fd7ac33-f8de-4a6b-8d45-0785e47f3db6)

This is happening before updates to URL Metrics:

![image](https://github.com/user-attachments/assets/5969e89b-27c9-46c3-b739-81328533771e)

hbhalodia commented:
Thanks, @westonruter For updating the PR and apologies that I was not available last friday, hence was not able to pick the further feedbacks.
